File: app/src/main/java/com/haivo/diana/Util/NoteUtils.java
package com.haivo.diana.Util;

import com.haivo.diana.Model.BaseNote;
import com.haivo.diana.Model.TunerOptions;

public final class NoteUtils {
    public static BaseNote.DetectedNote parseNote(String s, TunerOptions tunerOptions) {
        String note = s.substring(0, s.length() - 1);
        int octave = Integer.valueOf(s.substring(s.length() - 1, s.length()));
        double semitones = (octave - 4) * 12 + indexOfNote(BaseNote.Companion.getDEFAULT_NOTES(), note) - indexOfNote(BaseNote.Companion.getDEFAULT_NOTES(), "A");
        double BASE = tunerOptions.getTunerBase();
        double frequency = BASE * Math.pow(Math.pow(2, (1.00 / 12.00)), semitones);

        String translatedNote = tunerOptions.getNotes()[indexOfNote(BaseNote.Companion.getDEFAULT_NOTES(), note)];
        return new BaseNote.DetectedNote(translatedNote, octave, frequency, frequency, 0, tunerOptions);
    }

    public static BaseNote.DetectedNote parseNote(double frequency, TunerOptions tunerOptions) {
        double aboveA = Math.log((frequency / tunerOptions.getTunerBase())) / Math.log(Math.pow(2.00, 1 / 12.00));
        double closest = tunerOptions.getTunerBase() * Math.pow(Math.pow(2, 1.00 / 12.00), Math.round(aboveA));

        // NOTE ( 57 is the number of semitones of A4 )
        int index = (int) ((57 + Math.round(aboveA)) % 12);
        String note = tunerOptions.getNotes()[index];
        String localizedNote = BaseNote.Companion.getDEFAULT_NOTES()[index];

        // OCTAVE
        int octave = (int) (Math.floor((Math.log(frequency) - Math.log(tunerOptions.getTunerBase())) / Math.log(2) + 4.0));

        double offset = 100 * (aboveA - Math.round(aboveA));
        return new BaseNote.DetectedNote(note, octave, closest, frequency, offset, tunerOptions);
    }

    public static String[] getNotesByInstrument(BaseNote.Instrument mode) {
        switch (mode) {
            case KALIMBA:
                return BaseNote.Companion.getKALIMBA_NOTES();
            case FLUTE:
                return BaseNote.Companion.getVIETNAMESE_FLUTE_NOTES();
            default:
                return BaseNote.Companion.getCHROMATIC_NOTES();
        }
    }

    private static int indexOfNote(String[] a, String s) {
        int index = -1;
        for (int i = 0; i < a.length; i++) {
            if (a[i].equals(s)) {
                index = i;
                break;
            }
        }
        return index;
    }

    public static String getRealNote(TunerOptions tunerOptions, String translatedNote) {
        return BaseNote.Companion.getDEFAULT_NOTES()[getNoteIndex(tunerOptions, translatedNote)];
    }

    public static int getNoteIndex(TunerOptions tunerOptions, String translatedNote) {
        return indexOfNote(tunerOptions.getNotes(), translatedNote);
    }
}

File: app/src/main/java/com/haivo/diana/Util/NoteUtils.kt
package com.haivo.diana.Util

import com.haivo.diana.Model.BaseNote
import com.haivo.diana.Model.TunerOptions
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.pow
import kotlin.math.roundToInt

object NoteUtils {
    fun parseNote(s: String, tunerOptions: TunerOptions): BaseNote.DetectedNote {
        val note = s.substring(0, s.length - 1)
        val octave = Integer.valueOf(s.substring(s.length - 1, s.length))
        val semitones = ((octave - 4) * 12 + indexOfNote(BaseNote.DEFAULT_NOTES, note)
                - indexOfNote(BaseNote.DEFAULT_NOTES, "A")).toDouble()
        val base = tunerOptions.tunerBase.toDouble()
        val frequency = base * 2.0.pow(1.00 / 12.00).pow(semitones)

        val translatedNote = tunerOptions.notes[indexOfNote(BaseNote.DEFAULT_NOTES, note)]
        return BaseNote.DetectedNote(translatedNote, octave, frequency, frequency, 0.0, tunerOptions)
    }

    fun parseNote(frequency: Double, tunerOptions: TunerOptions): BaseNote.DetectedNote {
        val aboveA = ln(frequency / tunerOptions.tunerBase) / ln(2.00.pow(1 / 12.00))
        val closest = tunerOptions.tunerBase * 2.0.pow(1.00 / 12.00).pow(aboveA.roundToInt().toDouble())

        // NOTE ( 57 is the number of semitones of A4 )
        val index = ((57 + aboveA.roundToInt()) % 12).toInt()
        val note = tunerOptions.notes[index]
        val localizedNote = BaseNote.DEFAULT_NOTES[index]

        // OCTAVE
        val octave = floor((ln(frequency) - ln(tunerOptions.tunerBase.toDouble())) / ln(2.0) + 4.0).toInt()

        val offset = 100 * (aboveA - aboveA.roundToInt())
        return BaseNote.DetectedNote(note, octave, closest, frequency, offset, tunerOptions)
    }

    fun getNotesByInstrument(mode: BaseNote.Instrument): Array<String> {
        return when (mode) {
            BaseNote.Instrument.KALIMBA -> BaseNote.KALIMBA_NOTES
            BaseNote.Instrument.FLUTE -> BaseNote.VIETNAMESE_FLUTE_NOTES
            else -> BaseNote.CHROMATIC_NOTES
        }
    }

    private fun indexOfNote(a: Array<String>, s: String): Int {
        return a.indexOf(s)
    }

    fun getRealNote(tunerOptions: TunerOptions, translatedNote: String): String {
        return BaseNote.DEFAULT_NOTES[getNoteIndex(tunerOptions, translatedNote)]
    }

    fun getNoteIndex(tunerOptions: TunerOptions, translatedNote: String): Int {
        return indexOfNote(tunerOptions.notes, translatedNote)
    }
}
